File: app/src/main/java/com/arman/queuetube/model/adapters/BaseAdapter.java
package com.arman.queuetube.model.adapters;

import android.content.Context;
import android.view.ViewGroup;

import com.arman.queuetube.model.VideoData;
import com.arman.queuetube.model.viewholders.BaseViewHolder;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.function.Predicate;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public abstract class BaseAdapter<E, VH extends BaseViewHolder<E>> extends RecyclerView.Adapter<VH> {

    protected List<E> items;

    public BaseAdapter() {
        this(new ArrayList<E>());
    }

    public BaseAdapter(List<E> items) {
        this.items = items;
    }

    public E get(int index) {
        return this.items.get(index);
    }

    public boolean add(E item) {
        boolean res = this.items.add(item);
        if (res) {
            this.notifyItemInserted(this.items.size() - 1);
        }
        return res;
    }

    public boolean addAll(Collection<E> items) {
        int startRange = this.items.size();
        boolean res = this.items.addAll(items);
        if (res) {
            this.notifyItemRangeInserted(startRange, items.size());
        }
        return res;
    }

    public boolean remove(E item) {
        int index = this.items.indexOf(item);
        boolean res = this.items.remove(item);
        if (res) {
            this.notifyItemRemoved(index);
        }
        return res;
    }

    public E remove(int index) {
        E item = this.items.remove(index);
        if (item != null) {
            this.notifyItemRemoved(index);
        }
        return item;
    }

    public boolean removeAll(Collection<E> data) {
        boolean res = this.items.removeAll(data);
        this.notifyDataSetChanged();
        return res;
    }

    public boolean isEmpty() {
        return this.items.isEmpty();
    }

    public E pop() {
        return this.remove(0);
    }

    public void clear() {
        this.items.clear();
        this.notifyDataSetChanged();
    }

    @NonNull
    @Override
    public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void onBindViewHolder(@NonNull VH holder, int position) {
        holder.bind(this.items.get(position));
    }

    @Override
    public int getItemCount() {
        return this.size();
    }

    public boolean setAll(Collection<E> data) {
        this.items.clear();
        boolean ret = this.items.addAll(data);
        this.notifyDataSetChanged();
        return ret;
    }

    public int size() {
        return this.items.size();
    }

    public List<E> getAll() {
        return this.items;
    }

}

File: app/src/main/java/com/arman/queuetube/model/adapters/BaseAdapter.kt
package com.arman.queuetube.model.adapters

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.arman.queuetube.model.viewholders.BaseViewHolder
import java.util.*

abstract class BaseAdapter<E, VH : BaseViewHolder<E>> @JvmOverloads constructor(protected var items: MutableList<E> = ArrayList()) : RecyclerView.Adapter<VH>() {

    val isEmpty: Boolean
        get() = this.items.isEmpty()

    val all: List<E>
        get() = this.items

    operator fun get(index: Int): E {
        return this.items[index]
    }

    fun add(item: E): Boolean {
        val res = this.items.add(item)
        if (res) {
            this.notifyItemInserted(this.items.size - 1)
        }
        return res
    }

    fun addAll(items: Collection<E>): Boolean {
        val startRange = this.items.size
        val res = this.items.addAll(items)
        if (res) {
            this.notifyItemRangeInserted(startRange, items.size)
        }
        return res
    }

    fun remove(item: E): Boolean {
        val index = this.items.indexOf(item)
        val res = this.items.remove(item)
        if (res) {
            this.notifyItemRemoved(index)
        }
        return res
    }

    fun remove(index: Int): E? {
        val item = this.items.removeAt(index)
        if (item != null) {
            this.notifyItemRemoved(index)
        }
        return item
    }

    fun removeAll(data: Collection<E>): Boolean {
        val res = this.items.removeAll(data)
        this.notifyDataSetChanged()
        return res
    }

    fun pop(): E? {
        return this.remove(0)
    }

    fun clear() {
        this.items.clear()
        this.notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        throw UnsupportedOperationException()
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        holder.bind(this.items[position])
    }

    override fun getItemCount(): Int {
        return this.items.size
    }

    fun setAll(data: Collection<E>): Boolean {
        this.items.clear()
        val ret = this.items.addAll(data)
        this.notifyDataSetChanged()
        return ret
    }

}
